hack/make/.integration-test-helpers

```diff
@@ -1,65 +1,72 @@
 #!/usr/bin/env bash
-
-: ${TEST_REPEAT:=0}
-
-bundle_test_integration_cli() {
-	TESTFLAGS="$TESTFLAGS -check.v -check.timeout=${TIMEOUT} -test.timeout=360m"
-	go_test_dir integration-cli $DOCKER_INTEGRATION_TESTS_VERIFIED
-}
-
-# If $TESTFLAGS is set in the environment, it is passed as extra arguments to 'go test'.
-# You can use this to select certain tests to run, e.g.
-#
-#     TESTFLAGS='-test.run ^TestBuild$' ./hack/make.sh test-unit
 #
 # For integration-cli test, we use [gocheck](https://labix.org/gocheck), if you want
 # to run certain tests on your local host, you should run with command:
 #
-#     TESTFLAGS='-check.f DockerSuite.TestBuild*' ./hack/make.sh binary test-integration-cli
+#     TESTFLAGS='-check.f DockerSuite.TestBuild*' ./hack/make.sh binary test-integration
 #
-go_test_dir() {
-	dir=$1
-	precompiled=$2
-	testbinary="$ABS_DEST/test.main"
-	testcover=()
-	testcoverprofile=()
+
+source "$SCRIPTDIR/make/.go-autogen"
+
+: ${TEST_REPEAT:=1}
+
+integration_api_dirs=("$(
+	find ./integration -type d |
+	grep -vE '^(./integration$|./integration/util)')")
+
+run_test_integration() {
+	local flags="-test.v -test.timeout=${TIMEOUT} $TESTFLAGS"
+	for dir in $integration_api_dirs; do
+		(
+			cd $dir
+			echo "Running $PWD"
+			test_env ./test.main $flags
+		)
+	done
+
 	(
-		set -e
-		mkdir -p "$DEST/coverprofiles"
-		export DEST="$ABS_DEST" # in a subshell this is safe -- our integration-cli tests need DEST, and "cd" screws it up
-		if [ -z $precompiled ]; then
-			ensure_test_dir $1 $testbinary
-		fi
-		cd "$dir"
-		i=0
-		while ((++i)); do
-			test_env "$testbinary" $TESTFLAGS
-			if [ $i -gt "$TEST_REPEAT" ]; then
-				break
-			fi
-			echo "Repeating test ($i)"
-		done
+		flags="-check.v -check.timeout=${TIMEOUT} -test.timeout=360m $TESTFLAGS"
+		cd integration-cli > /dev/null
+		echo "Running $PWD"
+		test_env ./test.main $flags
 	)
 }
 
-ensure_test_dir() {
-	(
-		# make sure a test dir will compile
-		dir="$1"
-		out="$2"
-		echo Building test dir: "$dir"
-		set -xe
-		cd "$dir"
-		go test -c -o "$out" -ldflags "$LDFLAGS" "${BUILDFLAGS[@]}"
-	)
+build_test_suite_binaries() {
+	build_test_suite_binary ./integration-cli "test.main"
+	for dir in $integration_api_dirs; do
+		build_test_suite_binary "$dir" "test.main"
```

Comment:
> Does it make sense to build a binary for each package? 

I think the advantages out-weight the single disadvantage (disk space which is recovered when the test run ends).

> I imagine that each binary will be quite large, so this may take a lot of space

Yes, I think this is the only disadvantage. I don't think the disk space will be all that significant, but it will definitely grow as we add more packages. I expect it might take 200M or so once we have everything split into separate packages. If it becomes an issue we can add a flag to disable the build-first so they can be cleaned up after each package suite runs, assuming 200M actually becomes a problem for someone. Right now I don't think it's worth adding that complexity.

> and could be fairly slow

It shouldn't be any slower than running the complete test suite. `go test` builds this binary anyway, it just hides it. I guess when we add support for only running specific packages we will need a way to only build the binaries we are going to run. For now it's not any slower than not pre-building them, and it lets us fail fast when there are syntax errors, which is a big win.

> I believe we prebuild the integration-cli binary as a convenience so that syntax errors are caught before the dockerd binary is built.

Not before the `dockerd` binary is built, but before we run the slow test setup parts of starting the daemon and loading in fixtures.

> If there are going to be many binaries, maybe this doesn't make sense anymore, and it would be better to invoke tests with go test after the daemon is running.

If anything it's *more* important now. I don't want to run half the test suite then have it die because of a syntax error. It should build all the binaries first, then setup the environment, and finally run the tests. When we support running one package (or a subset of packages) at a time we can make sure it only builds the appropriate binaries. That feature should be easy to implement, just allow `integration_api_dirs` to be set by the user. 

> is there value in separating the integration tests into different packages if the script doesn't let you build/test packages individually?

Yes, I think there is a lot of value in this, it's one of the more important parts of this change. In the future, when we have more packakges, we can enable support for build/testing a single package or a subset of packages (as mentioned it should be really easy to add).

Until then I believe we still benefit from improved caching. As I understand it, if a package isn't modified go is able to use a cached version of it. By having more packages (as long as the packages are well separated) we know that all but one package should be cached on most runs. This should greatly reduce the startup time of the integration test suite (even without the further optimisation of running just a single package).